mat_dp_pipeline/pipeline/common.py

```diff
@@ -0,0 +1,135 @@
+
+        Raises:
+            ValueError: when validation fails
+
+        Returns:
+            set[str]: mismatched resources. It's up to the client code to decide
+            whether it's an error or not. No exception is thrown here if this set
+            isn't empty.
+        """
+        sdf.validate_tech_units(self.tech_metadata)
+        intensities_techs = self.intensities.index.droplevel(0).unique()
+        targets_techs = self.targets.index
+
+        validate(
+            set(targets_techs) <= set(intensities_techs),
+            "Target's techs must be a subset of intensities' techs!",
+        )
+
+        # move years to columns, reindex techs, bring the years back
+        self.intensities = (
+            self.intensities.unstack(level=0)  # type: ignore
+            .reindex(targets_techs)
+            .stack()
+            .reorder_levels(["Year", "Category", "Specific"])
+            .sort_index()
+        )
+        # unstack/stack messes with types
+        assert isinstance(self.intensities, pd.DataFrame)
+
+        intensities_resources = set(self.intensities.columns)
+        indicators_resources = set(self.indicators.index.get_level_values("Resource"))
+        common_resources = intensities_resources & indicators_resources
+        mismatched_resources = intensities_resources.symmetric_difference(
+            indicators_resources
+        )
+        if mismatched_resources:
+            sorted_resources = sorted(common_resources)
+            self.intensities = self.intensities.reindex(columns=sorted_resources)
+            self.indicators = self.indicators.reindex(
+                sorted_resources, level="Resource"
+            )
+        return mismatched_resources
+
+
+@dataclass(eq=False, order=False)
+class ProcessableInput:
+    """Single processable input. The frames in processable input do not contain year
+    dimension. This is the lowest level structure, ready for processing.
+
+    Attributes:
+        intensities (DataFrame): Tech x (Resource1, Resource2, ..., ResourceN)
+        targets (Series): Tech -> float. Tech keys must be a subset of intensities' keys
+        indicators (DataFrame): Resource x (Indicator1, Indicator2, ..., IndicatorM)
+            There should be exactly N resources, matching columns in intensities frame
+    """
+
+    intensities: pd.DataFrame
+    targets: pd.Series
+    indicators: pd.DataFrame
+
+    def copy(self) -> "ProcessableInput":
+        return ProcessableInput(
+            intensities=self.intensities.copy(),
+            targets=self.targets.copy(),
+            indicators=self.indicators.copy(),
+        )
+
+    def save(self, directory: Path, exist_ok: bool = False) -> None:
+        """Save ProcessableInput to files in a directory
+
+        Args:
+            directory (Path): output directory (must exist)
+            exist_ok (bool, optional):  Is it OK if files exist already. They will be overriden if so. Defaults to False.
```

Comment:
overridden

Reply:
exist_ok -> exists?

Reply:
why?

Reply:
Oh, exists_ok ?

Reply:
Not sure I understand. Please clarify.

Reply:
Just a weird name

Reply:
`Is it OK if files exist already` - this is good explanation I suppose

Reply:
it's also called exist_ok not exists_ok - maybe a typo?

Reply:
"it exists, OK??" haha

Reply:
https://docs.python.org/dev/library/os.html#os.makedirs , https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir